bot/utils/helpers.py

```diff
@@ -6,3 +6,30 @@ def suppress_links(message: str) -> str:
     for link in set(re.findall(r"https?://[^\s]+", message, re.IGNORECASE)):
         message = message.replace(link, f"<{link}>")
     return message
+
+def neutralise_string(txt: str | None) -> str | None:
+    """Attempts to neutralise all punctuation and cases and returns a string of lowercase words."""
+    # return early if no text provided.
+    if not txt:
+        return None
+
+    # take out punctuation
+    txt = re.sub(r"([^\w\s]|_)", " ", txt)
+
+    # full caps words but leaves camelCase / PascalCase
+    words = [word.lower() if word.isupper() else word for word in txt.split()]
+    txt = " ".join(words)
+
+    # attempt to split PascalCase and camelCase
+    words = []
+    old_i = 0
+    for i, char in enumerate(txt):
+        # to avoid PascalCase getting leading empty append
+        if char.isupper() and i != 0:
+            words.append(txt[old_i:i])
+            old_i = i
+    words.append(txt[old_i:])
```

Comment:
There's quite a lot of potentially excessive `" ".join` and `.split()`s here.

Especially if you're going to opt for the solution in my first review comment, we're doing a `.split` immediately after a `" ".join` when `neutralise_string` returns.

One way we could improve this is to iterate over each word using an outer for-loop, before looping through each individual word like on line 26, and append to something like a `new_words` list. This removes the need of the `.join` call on line 21.

This then gives us the opportunity to repurpose the helper function to return the list of words instead, since each word in the list would no longer have spaces in them. This removes the need for `.join` when returning from the function and the subsequent `.split` in each of the commands.

Again, this suggestion is also up to you on whether you'd like to eliminate the repeated `.join` and `.split` calls, since one might argue that the function becomes longer or less readable just to save these extra calls.